### PR TITLE
Reimplement with zip and two DirWalkers

### PR DESCRIPTION
**Pull request by mikerite**
The current implementation has the problem that every matching file is
opened and read twice. This is an alternate implementation that only checks
every pair of files once.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -33,68 +34,36 @@ pub enum Error {
...
-
-        if a_text != b_text {
-            return Ok(true);
-        }
-    }
+fn walk_dir<P: AsRef<Path>>(path: P) -> std::iter::Skip<walkdir::IntoIter> {
+    WalkDir::new(path)
+        .sort_by(compare_by_file_name)
+        .into_iter()
+        .skip(1)
```
Comment (epage) [inline src/lib.rs]: btw we didn't need to `skip` in the original implementation.  Why is this needed now?

File: src/lib.rs [hunk 2]
```diff
@@ -33,68 +34,36 @@ pub enum Error {
...
+    for (a, b) in (&mut a_walker).zip(&mut b_walker) {
+        let a = a?;
+        let b = b?;
 
-    let a_count = std::fs::read_dir(&a_base)?.count();
-    let b_count = std::fs::read_dir(&b_base)?.count();
-
-    if a_count != b_count {
-        return Ok(true);
+        if a.file_type() != b.file_type() || a.file_name() != b.file_name()
```
Comment (epage) [inline src/lib.rs]: This is checking just `file_name`.  I feel I'm missing something.  In the case of diffing:
- a_base/foo/file.txt
- a_base/bar/file.txt
It looks like this will say there is no difference when in fact there is.

With only one file in each base, the sort won't help us spot a discrepancy.  We'' then check the file name and file contents which will be the same but the directories are different.

Do we instead need to check `a.strip_prefix(a_base) != b.strip_prefix(b_base)`?

File: src/lib.rs (hunk 1 above)
Comment (mikerite) [reply src/lib.rs]: WalkDir starts iteration with the path you give it. If you take the `skip` out, the tests will fail when it compares 'a_base' to 'b_base'. In the old code this wasn't necessary since every comparison would start with the pointless check that ('a_base').strip_prexfix('a_base') exists in 'b_base'.

File: src/lib.rs (hunk 2 above)
Comment (mikerite) [reply src/lib.rs]: I've corrected the error. `strip_prefix` is unnecessary it's enough to check depth.

File: tests/smoke.rs [hunk 3]
```diff
@@ -33,4 +37,24 @@ fn oneempty() {
 #[test]
 fn reflexive() {
     assert!(dir_diff::is_different("tests/reflexive/dir1", "tests/reflexive/dir2").unwrap());
-}
\ No newline at end of file
+}
+
+fn ensure_dir<P: AsRef<Path>>(path: P) -> Result<(), std::io::Error> {
+    match create_dir(path) {
```
Comment (epage) [inline tests/smoke.rs]: How come this is needed?

File: tests/smoke.rs [hunk 4]
```diff
@@ -33,4 +37,24 @@ fn oneempty() {
...
+        Err(ref err) if err.kind() == ErrorKind::AlreadyExists => Ok(()),
+        other => other,
+    }
+}
+
+#[test]
+fn filedepth() {
```
Comment (epage) [inline tests/smoke.rs]: This is testing the case of `dir/file.txt` vs `file.txt` but what about `dirA/file.txt` vs `dirB/file.txt`?

File: src/lib.rs (hunk 1 above)
Comment (epage) [reply src/lib.rs]: Ok, so its an optimization to avoid checking the current dir.  Does this deserve a comment?

Comment (epage) [reply src/lib.rs]: (not for the what but the why)

File: tests/smoke.rs (hunk 4 above)
Comment (mikerite) [reply tests/smoke.rs]: This was never a problem. The walkers will compare `dirA` and `dirB` and the difference will be detected. I added a test for this to be sure.

File: tests/smoke.rs (hunk 3 above)
Comment (mikerite) [reply tests/smoke.rs]: It's needed to create an empty directory for the test because it's not possible to add an empty directory to git. The work around is add a ".gitkeep" file inside the empty directory but then it's no longer empty.

File: src/lib.rs (hunk 1 above)
Comment (mikerite) [reply src/lib.rs]: It's not an optimization. You need to skip the base dir in the comparison, otherwise two dirs will always be treated as different unless they have the same name.

Comment (epage) [reply src/lib.rs]: You're right; thanks for pointing that out!

File: tests/smoke.rs (hunk 4 above)
Comment (epage) [reply tests/smoke.rs]: Ah, been a while since I've messed enough with WalkDir.  Thanks for pointing that out.

File: tests/smoke.rs (hunk 3 above)
Comment (epage) [reply tests/smoke.rs]: I was more so thinking higher level (with the test data split, it was harder to see whats going on).

So it looks like you have
```
asc/dir1/a/b.txt
asc/dir2/a
asc/dir2/b.txt
```
and
```
desc/dir1/a.txt
desc/dir1/b
desc/dir2/b/a.txt
```

So this will ensure that the entries for the `a` and `b` directories are the same, so we can progress to compare the `a.txt` and `b.txt` files, ensuring that they are treated differently.

That right?

**Comment by epage** [comment]
Now that reviewing is done for reals this time, could you squash your commits?

**Comment by mikerite** [comment]
I squashed the commits in a new branch. The new pull request is #14.
